crates/rpc/src/server/api.rs

```diff
@@ -40,28 +42,29 @@ type StoreClient = store_client::ApiClient<InterceptedService<Channel, OtelInter
 type BlockProducerClient =
     block_producer_client::ApiClient<InterceptedService<Channel, OtelInterceptor>>;
 
-pub struct RpcApi {
+pub struct RpcService {
     store: StoreClient,
     block_producer: BlockProducerClient,
 }
 
-impl RpcApi {
-    pub(super) async fn from_config(config: &RpcConfig) -> Result<Self, Error> {
-        let channel = tonic::transport::Endpoint::try_from(config.store_url.to_string())?
-            .connect()
-            .await?;
+impl RpcService {
+    pub(super) async fn new(
+        store_address: SocketAddr,
+        block_producer_address: SocketAddr,
+    ) -> Result<Self, Error> {
+        let store_url = format!("http://{store_address}");
+        let channel = tonic::transport::Endpoint::try_from(store_url)?.connect().await?;
         let store = store_client::ApiClient::with_interceptor(channel, OtelInterceptor);
-        info!(target: COMPONENT, store_endpoint = config.store_url.as_str(), "Store client initialized");
+        info!(target: COMPONENT, store_endpoint = %store_address, "Store client initialized");
 
-        let channel = tonic::transport::Endpoint::try_from(config.block_producer_url.to_string())?
-            .connect()
-            .await?;
+        let block_producer_url = format!("http://{block_producer_address}");
+        let channel = tonic::transport::Endpoint::try_from(block_producer_url)?.connect().await?;
```

Comment:
This is a bit awkward and makes me think I was too hasty in removing our custom `Endpoint` type in #654.

As a summary:

- Tonic wants a `Uri` but accepts strings in the format of `http(s)://{SocketAddr}`.
- `TcpListener::bind` wants a `SocketAddr`.
- `Uri` cannot be trivially converted to a `SocketAddr`
- `Url` can be converted to a `SocketAddr` so we use it.
- Clients may want to perform a DNS lookup to the server address of `http://rpc.testnet.miden.io`

I think its okay to leave as is, but we probably want to revisit this (again :'( ).

Reply:
I don't actually know. It just seems messy and incorrect to me at the moment.

As an example, if you provide an `https` URL for a server at the moment then we don't actually support that. We strip the protocol from the URL when we convert to a socket address and don't add TLS support when we create the tonic server.

This is sort of fine because we don't (ourselves) want to use TLS internally. But it sort of points out the leaky abstraction here.

If we implement something ourselves then we'll probably have the opposite issue of rejecting weird but valid URLs.

Reply:
Actually maybe the real answer is making startup robust. The only reason we have `TcpListener` involved is to separate server initialization from the clients spinning up in `serve`.

If we have the components capable of starting up robustly with internal retries etc, then we can simply use `Uri` which is what `tonic` can use.